### ISSUE TITLE
Add Environment Markers to pypi-dependencies specification

### ISSUE THREAD

**Issue by tdejager** [prefix-dev/pixi#1259]
### Problem description

In pypi packaging you have the concept of environment markers, which can help you to optionally install a dependency. We've had requests of people wanting to use this to exclude certain dependencies according to the python version e.g:

```toml
pkg = { path ="./my-pkg", env-markers = ["python_version < '3.11'] }
```
Because we re-create the PEP508 requirements, it would be easy to add and this is already taken into account in the satisfyiability. The main upside would be less features when compared to splitting all this up. Also, it would not be very hard to integrate.

Would love to hear everyone's thoughts.

## Previous discussions

We have also talked about explicitly excluding certain dependencies from environments. However, @baszalmstra and myself feel this would open a can of worms. Seeing as how long it's taking to even agree on how to exclude the default-feature: https://github.com/prefix-dev/pixi/pull/1092 this approach mentioned above would be easier and already supported by the pypi ecosystem.

This kind of leaves the same question for conda dependencies, but I think we might need to tackle this differently, although a similar feature would probably be desirable.

**Comment by abey79** [comment]
I found myself needing this again today, while trying to add proper support for testing against various python version in Rerun.

**Comment by deltamarnix** [comment]
I have a pyproject.toml that contains a number of optional dependencies, based on the python version and system you are using:

```toml
optional = [
    "affine",
    "descartes",
    "fiona",
    "geojson",
    "geopandas",
    "imageio",
    "netcdf4",
    "pymetis ; platform_system != 'Windows'",
    "pyproj",
    "pyshp",
    "pyvista ; python_version <'3.13'",
    "rasterio",
    "rasterstats",
    "scipy",
    "shapely >=2.0",
    "vtk ; python_version <'3.13'",
    "xmipy",
]
```

My workaround is now to cap the python version and overwrite pymetis with a conda dependency, because there is a windows version on conda. But it's not the same functionality in some way or another.

```toml
[pixi.tool.feature.optional.dependencies]
pymetis = "*"
```

**Comment by tdejager** [comment]
Yeah we are in the process of doing a uv upgrade and are making a distinction between the open-source `pep_*` crates and the `uv_pep` crates, now we should be able to integrate environment markers more easily :)

**Comment by cpcloud** [comment]
Would really love to see environment marker support in Pixi. Just out of curiosity is that mentioned uv upgrade complete?

**Comment by tdejager** [comment]
It has been completed indeed!

**Comment by tdejager** [comment]
Just added this to start on into our cycle in two weeks :)

**Comment by hameerabbasi** [comment]
If this is something I can help with, I'd be happy to, if I'm guided on what code to touch. 😄

**Comment by tdejager** [comment]
For sure! Let me write something up tomorrow. I was waiting for the switch to the crates.io uv crates in rattler. But does not seem to be happening soon, so let's fix it first!

**Comment by tdejager** [comment]
Okay, so the problem is we have a conversion issue between the open-source crates and the uv git crates. So I made a repo with claude once if we can do like a string <-> string conversion, and if all cases are supported. My conclusion is that this is possible: https://github.com/tdejager/pep-env-markers-demo (this was vibecoded). 

So I think the steps need to be,

1. Make sure we have a pypi-dependencies specification for it, so it can be used in `pixi.toml`.
2. Parse both these and the pypi requirement from `pyproject.toml` as a `pep_508rs` type.
3. Wire up conversion methods to the uv-types in `uv_conversions` crate in the pixi project.
4. Check if `rattler_lock` needs support for it? I think it already does though.
5. Convert to the uv env marker *before* the solve.
6. Convert back to `pep508_rs` *after* the solve.

The whole thing needs tests for most of these points

Feel free to use an agent from my part, but make sure to make an extensive plan, have it check it for holes beforehand and maybe split the PR's if possible e.g: first the specification (with warnings that it is unused), then conversion and solving. Makes sense?

**Comment by hameerabbasi** [comment]
Makes sense; although I don't vibe code usually. I'll read the relevant bits of code and might hit you up on Discord with questions. Thanks for the detailed plan!
